os/kill.go

```diff
@@ -61,27 +74,123 @@ func getAllChildren(proc *process.Process) []*process.Process {
+func wrapWithStopAndCont(signal syscall.Signal) []syscall.Signal {
 	signals := []syscall.Signal{syscall.SIGSTOP, signal}
 	if signal != syscall.SIGKILL { // no point in sending any signal after SIGKILL
 		signals = append(signals, syscall.SIGCONT)
 	}
+	return signals
+}
 
-	for _, currentSignal := range signals {
-		if err := sendSignalToProcessGroups(currentSignal, pgids); err != nil {
-			return err
+func sendSignalsToProcessGroups(signals []syscall.Signal, pgids []int) error {
+	for _, signal := range signals {
+		for _, pgid := range pgids {
+			log.Infof("Sending signal %s to pgid %d", signal, pgid)
+			err := syscall.Kill(-pgid, signal)
+			if err != nil {
+				log.Infof("Error sending signal to pgid %d: %s", pgid, err)
+			}
 		}
 	}
 	return nil
 }
 
-func sendSignalToProcessGroups(signal syscall.Signal, pgids []int) error {
+// KillTreeWithExcludes sends signal to whole process tree, starting from given pid as root.
+// Omits processes matching names specified in processesToExclude. Kills using pids instead of pgids.
+func KillTreeWithExcludes(signal syscall.Signal, pid int32, processesToExclude []string) error {
+	log.Infof("Will send signal %s to tree starting from %d", signal.String(), pid)
+
+	pgids, err := getProcessGroupsInTree(pid)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Found process groups: %v", pgids)
+
+	pids, err := findProcessesInGroups(pgids)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Found processes in groups: %v", pids)
+
+	pids, err = excludeProcesses(pids, processesToExclude)
+	if err != nil {
+		return err
+	}
+
+	signals := wrapWithStopAndCont(signal)
+	return sendSignalsToProcesses(signals, pids)
+}
+
+func findProcessesInGroups(pgids []int) ([]int, error) {
+	var pids []int
 	for _, pgid := range pgids {
-		log.Infof("Sending signal %s to pgid %d", signal, pgid)
-		err := syscall.Kill(-pgid, signal)
+		cmd := exec.Command("pgrep", "-g", strconv.Itoa(pgid))
+		output, err := cmd.CombinedOutput()
 		if err != nil {
-			log.Infof("Error sending signal to pgid %d: %s", pgid, err)
-			return err
+			return nil, fmt.Errorf("'pgrep -g %d' failed: %s", pgid, err)
+		}
+		if !cmd.ProcessState.Success() {
+			return nil, fmt.Errorf("'pgrep -g %d' failed, output was: '%s'", pgid, output)
+		}
+
+		scanner := bufio.NewScanner(strings.NewReader(string(output)))
+		for scanner.Scan() {
+			pid, err := strconv.Atoi(scanner.Text())
+			if err != nil {
+				return nil, fmt.Errorf("cannot convert pgrep output: %s. Output was '%s'", err, output)
+			}
+			pids = append(pids, pid)
+		}
+	}
+
+	return pids, nil
+}
+
+func excludeProcesses(pids []int, processesToExclude []string) ([]int, error) {
+	var retainedPids []int
+	for _, pid := range pids {
+		proc, err := process.NewProcess(int32(pid))
+		if err != nil {
+			return nil, err
+		}
+
+		name, err := proc.Name()
+		if err != nil {
+			log.Infof("Could not get process name of %d, continuing", pid)
+			continue
```

Comment:
every process that isn't on processesToExclude list should be terminated by default settings. 
-> Could not get process name of %d not excluding.